environ.go

```diff
@@ -0,0 +1,304 @@
+// Package environ provides an easy way to store and manipulate environment state
+// to provide to other packages like exec, where you want to limit the effect of
+// calling environment on shells spawned by your application.
+package environ
+
+import (
+	"encoding/json"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// An Environ holds a set of environment variables for manipulation.
+type Environ struct {
+	l locker
+	m map[string]string
+}
+
+// FromOS returns an Environ containing the current os.Environ().
+func FromOS() *Environ {
+	return New(os.Environ())
+}
+
+// Len returns the length of the underling environment map.
+func (e *Environ) Len() int {
+	defer e.readLocker()()
+
+	return len(e.m)
+}
+
+// MarshalJSON satisfies json.Marshaler interface.
+func (e *Environ) MarshalJSON() ([]byte, error) {
+	defer e.readLocker()()
+
+	return json.Marshal(e.AsSlice())
+}
+
+// UnmarshalJSON satisfies json.Unmarshaler interface.
+func (e *Environ) UnmarshalJSON(data []byte) error {
+	var environ []string
+	if err := json.Unmarshal(data, &environ); err != nil {
+		return err
+	}
+
+	*e = *New(environ)
+
+	return nil
+}
+
+// New creates an Environ from a list of "key=value" strings.
+func New(environ []string) *Environ {
+	return &Environ{
+		l: new(sync.RWMutex),
+		m: envSliceAsMap(environ),
+	}
+}
+
+// Set updates the Environ, replacing the value at key with val. If
+// such already exists, it'll be clobbered.
+func (e *Environ) Set(key, val string) {
+	defer e.writeLocker()()
+
+	e.m[key] = val
+}
+
+// Unset deletes key's value from the Environ.
+func (e *Environ) Unset(key string) {
+	defer e.writeLocker()()
+
+	delete(e.m, key)
+}
+
+// Get retrieves the value in the Environ under key, or "" if missing.
+func (e *Environ) Get(key string) string {
+	defer e.readLocker()()
+
+	return e.m[key]
+}
+
+// Keep scans the Environ looking for keys matching the keepers slice and
+// keeps them while discarding all others.
+//
+// It returns the slice of keys it could not find, because this may be a
+// condition for failure in some situations. If a key in the keepers list
+// has the suffix "*", it will use wild-card logic to capture columns.
+func (e *Environ) Keep(keepers ...string) []string {
+	m := e.AsMap()
+	missing := keep(&m, keepers)
+
+	defer e.writeLocker()()
+
+	e.m = m
+
+	return missing
+}
+
+func keep(m *map[string]string, keepers []string) (missing []string) {
+	missing = make([]string, 0, len(keepers))
+	keeping := make(map[string]string, len(keepers))
+
+	for _, keeperKey := range keepers {
+		var found bool
+
+		// Matching on wildcards
+		if strings.HasSuffix(keeperKey, "*") {
```

Comment:
what about using one of the globbing packages? that would give us more globbing functionality around like what about if glob was `*_TOKEN` or `AWS_*_KEY`

Reply:
Originally, I was going to search for a match on `"[-_]"` to limit it to complete matches. The real use case is applications with their own prefixes, not expansions on globs.

Reply:
I don't know why you'd use `*_TOKEN` instead of the *exact* token you need to feed into the environment. I also think you'd just want `AWS_*` because the domain is more important they type of value (in this case `_KEY`)

Reply:
i'm thinking more about drop than keep - eg I currently have GHE_TOKEN, GITHUB_TOKEN, JIRA_API_TOKEN, and METWORX_API_TOKEN in my environ, so from a sanitization perspective the situation of, I want to match the global env a bit but want to just put some protection is, especially thinking in terms of this in general purpose settings. 

That said, its such a small thing to add if we ever have an actual demonstrated use case, this likely will cover the vast majority of cases, and worst case a higher level implementation could still just glob themselves and pass in the end results.

My one request would then be, can we update the readme to reflect its just a prefix glob (I will add suggestion)

Reply:
I'm just going to switch to using a regexp and returning err if it doesn't compile.